cpp/turbodbc_arrow/Library/src/set_arrow_parameters.cpp

```diff
@@ -0,0 +1,375 @@
+    };
+
+    std::vector<std::unique_ptr<parameter_converter>> make_converters(
+        Table const & table,
+        turbodbc::bound_parameter_set & parameters)
+    {
+        std::vector<std::unique_ptr<parameter_converter>> converters;
+
+        for (int64_t i = 0; i < table.num_columns(); ++i) {
+            std::shared_ptr<ChunkedArray> const & data = table.column(i)->data();
+            arrow::Type::type dtype = data->type()->id();
+
+            switch (dtype) {
+              case arrow::Type::NA:
+                converters.emplace_back(new null_converter(data, parameters, i));
+              case arrow::Type::INT64:
+                converters.emplace_back(new int64_converter(data, parameters, i));
+                break;
+              case arrow::Type::BINARY:
+              case arrow::Type::STRING:
+                converters.emplace_back(new string_converter(data, parameters, i));
+                break;
+              case arrow::Type::DATE32:
+                converters.emplace_back(new date_converter(data, parameters, i));
+                break;
+              case arrow::Type::TIMESTAMP:
+                {
+                  auto const& time_dtype = static_cast<TimestampType const&>(*data->type());
+                  switch (time_dtype.unit()) {
+                    case TimeUnit::MICRO:
+                      converters.emplace_back(new microsecond_converter(data, parameters, i));
+                      break;
+                    case TimeUnit::NANO:
+                      converters.emplace_back(new nanosecond_converter(data, parameters, i));
+                      break;
+                    default:
+                      throw turbodbc::interface_error("Unsupported timestamp resolution");
+                  }
+                }
+                break;
+              case arrow::Type::BOOL:
+                converters.emplace_back(new bool_converter(data, parameters, i));
+                break;
+              case arrow::Type::DOUBLE:
+                converters.emplace_back(new double_converter(data, parameters, i));
+                break;
+              default:
+                std::ostringstream message;
+                message << "Unsupported Arrow type for column " << (i + 1) << " of ";
+                message << table.num_columns() << " (" << data->type()->ToString() << ")";
+                throw turbodbc::interface_error(message.str());
+            }
+        }
+
+        return converters;
+    }
+}
+
+void set_arrow_parameters(turbodbc::bound_parameter_set & parameters, pybind11::object const & pyarrow_table) {
+  arrow::py::import_pyarrow();
+  if (arrow::py::is_table(pyarrow_table.ptr())) {
+    std::shared_ptr<Table> table;
+    if (not arrow::py::unwrap_table(pyarrow_table.ptr(), &table).ok()) {
+      throw turbodbc::interface_error("Could not unwrap the C++ object from Python pyarrow.Table");
+    }
+
+    if (static_cast<int32_t>(parameters.number_of_parameters()) != table->num_columns()) {
```

Comment:
please extract this into a function like `assert_table_columns_match_parameters()` or similar